README.md
# GPR Data Visualizer

The GPR Data Visualizer is a Python-based research software for reading DZT format binary files through a simple graphical user interface (GUI).
The software is capabable of reading multiple DZT format binary files, applying various pre-defined filters to converted data, plotting the resulting data as human-readable plots using [matplotlib](https://matplotlib.org/), and exporting resulting data into human-readable formats.

##Installation

### Installation for Systems with Multi-User Anaconda3 Installations on Windows ###

Instructions for installing Anacond3 for Multiple Users can be found here [Multiple User Anaconda Installation](https://docs.anaconda.com/anaconda/install/multi-user/) Find refernce info to Anaconda in the [Documentation](https://github.com/GPR-Data-Visualizer/GPR-Data-Visualizer/blob/main/DOCS/GPR%20Documentation.pdf)

Installations of Anaconda3 of this type must use the default installation path:
```bash
C:\ProgramData\Anaconda3
```
Download the repository as a ZIP folder.

Extract to any location on your system.

Run [Public Installer](https://github.com/GPR-Data-Visualizer/GPR-Data-Visualizer/blob/main/Public_Install_GPR_Visualizer.bat) as administrator. GPR Data Visualizer's installation location will be:
```bash
C:\ProgramData\gpr_data_vis
```
### Installation for Systems with Single-User Anaconda3 Installations on Windows ###

Instructions for installing Anacond3 for a Single User can be found here [Single User Anaconda Installation](https://docs.anaconda.com/anaconda/install/windows/)

Installations of Anaconda3 of this type should use the default installation path to guarantee proper installation of GPR Data Visualizer:
```bash
C:\Users\Username\anaconda3
```
Download the repository as a ZIP folder.

Extract to any location on your system.


Run [Private Installer](https://github.com/GPR-Data-Visualizer/GPR-Data-Visualizer/blob/main/Private_Install_GPR_Visualizer.bat) as administrator. Assuming user possesses default homedrive and homepath system variables, GPR Data Visualizer's installation location will be:
```bash
C:\Users\Username\gpr_data_vis
```

## Usage

The installation process will generate the Desktop Shortcut GPR_Data_Vis which can be double-clicked to run the program.
For further usage instructions, see the documetation: [Documentation](https://github.com/GPR-Data-Visualizer/GPR-Data-Visualizer/blob/main/DOCS/GPR%20Documentation.pdf).

## License
[GNU GPL v3](https://www.gnu.org/licenses/gpl-3.0.en.html)
<!--
**GPR-Data-Visualizer/GPR-Data-Visualizer** is a ✨ _special_ ✨ repository because its `README.md` (this file) appears on your GitHub profile.

Here are some ideas to get you started:

- 🔭 I’m currently working on ...
- 🌱 I’m currently learning ...
- 👯 I’m looking to collaborate on ...
- 🤔 I’m looking for help with ...
- 💬 Ask me about ...
- 📫 How to reach me: ...
- 😄 Pronouns: ...
- ⚡ Fun fact: ...
-->


readgssi/CHANGES.md
# Changelog

## changes since 0.0.17
- fixed the read of the byte that contains `rh_system` and `rh_version` in the header ([#26](https://github.com/iannesbitt/readgssi/issues/26))
- fixed sampling depth interpretation ([#28](https://github.com/iannesbitt/readgssi/issues/28))
- added DZT file output ([#29](https://github.com/iannesbitt/readgssi/pull/29))
- added documentation for displaying absolute value of down-trace gradient in plots
- fixed a bug that incorrectly displayed long axis units on stacked arrays
- added [conda installation method](https://readgssi.readthedocs.io/en/latest/installing.html#installing-from-anaconda-cloud) and streamlined installation documentation
- switched version handling to file (`_version.py`)
- fixed a bug that did not handle `rhf_depth == 0` cases correctly
- added support for the 3102 antenna (500 MHz)
- fixed a bug that incorrectly plotted TWTT
- updated examples in Readme and fixed the autogen script (`make_plots.sh`)
- corrected an error with how `setup.py` got the version number, which prevented the creation of a conda skeleton

## changes since 0.0.16
- merged [#19](https://github.com/iannesbitt/readgssi/pull/19) from [teshaw](https://github.com/teshaw) which allows the user to specify partial read parameters for DZT files
- reintroduced a testing feature which creates CSV files from DZG records, in order to easily import location information to GIS
- added support for the StructureScan Mini XT which suffered from the unfortunate fate of having an antenna named 'none'
- improved handling of multiple user-specified frequencies
- adding the ability to plot the absolute value of the down-trace gradient, which can increase faint layer visibility at the cost of wave front resolution
- now use `rf_top` to correct depths, a major bug in previous versions
- limited distance normalization console output to one message every ten chunks (in the future, this will be controlled with a \r statement)
- added pause correction function to fix DZG/DZT trace number offset when collection is paused and GPS is active
- added ability to plot user marks as vertical lines (similar to the output on the controller itself)
- corrected module call documentation for plotting ([#23](https://github.com/iannesbitt/readgssi/issues/23))

## changes since 0.0.15
- fixed [#14](https://github.com/iannesbitt/readgssi/issues/14) which incorrectly asserted that user-set samples per meter was negative when it was not (sign was flipped) but processed the file correctly regardless
- moved zoom calculation and automatic file naming to their own functions
- added sphinx autodoc descriptions to some functions, in order to work towards solving [#9](https://github.com/iannesbitt/readgssi/issues/9)
- added comments to datetime parsing function
- fixed sampling frequency calculations to the value of `epsr` set in the header, which will prevent perceived frequency wander in frequency-based filters if the user sets `epsr`
- changed the behavior of the readdzt script to automatically search for a DZG file of the same name, and load that data if possible. this should allow python users to read all three variables (header, arrays, gps) if GPS exists.
- fixed command line control of DPI
- switched to "gray" colormap from "Greys". "Greys" rendered opposite polarity [#18](https://github.com/iannesbitt/readgssi/issues/18)

## changes since 0.0.14
- added support for the 5103 400 MHz antenna, and an antenna called `CUSTOM` in the header, for which center frequency is not known but is likely adjustable
- added support for the 350 HS hyperstacking antenna
- fixed a bug that created an extra matplotlib axis
- added rounding to epsilon_r value on y-axis depth-type display
- added a patch for [#12](https://github.com/iannesbitt/readgssi/issues/12) to improve axis sizing and tight_layout() handling
- changed background removal (BGR) function to operate similar to RADAN boxcar function
  - this change means that you can no longer simply specify `-r` for full-width background removal. new nomenclature for full-width BGR flag is `-r 0`. `-r 100` will apply a 100-trace rolling boxcar BGR (after normalization and stacking)
  - this is done with the `scipy.ndimage.filters.uniform_filter1d()` function and means that `scipy` is now explicitly required
  - function will automatically filter out low-frequency washout noise caused by bgr, using obspy's implementation of scipy's butterworth bandpass from `0.1*f` to `10*f`, where `f` is antenna frequency. nyquist is automatically accounted for by obspy.
  - BGR value is displayed in title if set > 0, or as "full" if BGR is full-width (i.e. `-r 0`)
- added zerophase vertical triangular FIR filter, and changed default filter from butterworth to triangular to attempt to match that in RADAN. butterworth is no longer used (but kept for posterity).
  - this function uses the `scipy.signal.firwin()` filter design and `scipy.signal.lfilter()` implementation
- added ability to suppress plot title using `-T` or `--titleoff` from the command line or title=False in main python function
- changed `ns_per_zsample` in header to read data shape from header rather than array size
- added a bug fix for filtering that caused the sample rate to be set too high
- significantly improved triangular bandpass filter, so that it nearly matches RADAN
- when Z axis value is set to "samples", the zeroed samples are taken into account. so, if the array had 150 samples sliced off, the plot axis will now start at 150 instead of 0
- added per-channel handling of timezero. you can pass a list with `-Z 233,150,92` (or `zero=[233,150,92,None]` in python). defaults to `-Z 2` (`zero=[2,None,None,None]`)
- added programmable zoom with `-e left,right,up,down` flag and list
- cleaned up some dangling floats
- updated help text and README help text
- added automatic tight bbox axes for savefig

readgssi/readgssi/constants.py
import pytz

"""
This module contains a number of variables that readgssi needs to perform physics calculations and interpret files from DZT files.
"""

MINHEADSIZE = 1024 # absolute minimum total header size
PAREASIZE = 128 # fixed info header size
RGPSSIZE = 9 # GPS record size
GPSAREASIZE = RGPSSIZE * 2
INFOAREASIZE = MINHEADSIZE - PAREASIZE - GPSAREASIZE

TZ = pytz.timezone('UTC')

# some physical constants for Maxwell's equation for speed of light in a dielectric medium
C = 299792458                   # speed of light in a vacuum
Eps_0 = 8.8541878 * 10**(-12)   # epsilon naught (vacuum permittivity)
Mu_0 = 1.257 * 10**(-6)         # mu naught (vacuum permeability)


# the GSSI field unit used
UNIT = {
    0: 'synthetic/gprMax',
    1: 'unknown system type',
    2: 'SIR 2000',
    3: 'SIR 3000',
    4: 'TerraVision',
    6: 'SIR 20',
    7: 'StructureScan Mini',
    8: 'SIR 4000',
    9: 'SIR 30',
    10: 'unknown system type',
    11: 'unknown system type',
    12: 'UtilityScan DF',
    13: 'HS',
    14: 'StructureScan Mini XT',
}

# a dictionary of standard gssi antenna codes and frequencies
# unsure of what they all look like in code, however
ANT = {
    '100MHz': 100,
    '200MHz': 200,
    '270MHz': 270,
    '350MHz': 350,
    '400MHz': 400,
    '500MHz': 500,
    '800MHz': 800,
    '900MHz': 900,
    '1600MHz': 1600,
    '2000MHz': 2000,
    '2300MHz': 2300,
    '2600MHz': 2600,
    '3200': 'adjustable',
    '3200MLF': 'adjustable',
    'gprMa': 'adjustable',      # gprMax support
    'GSSI': 'adjustable',       # support for issue #11
    'CUSTOM': 'adjustable',
    '3207': 100,
    '3207AP': 100,
    '5106': 200,
    '5106A': 200,
    '50300': 300,
    '350': 350,
    '350HS': 350,
    'D400HS': 350,
    '50270': 270,
    '50270S': 270,
    'D50300': 300,
    '5103': 400,
    '5103A': 400,
    '50400': 400,
    '50400S': 400,
    '3102 500MHZ': 500,
    '3102': 500, # presumed to exist
    '800': 800,
    'D50800': 800,
    '3101': 900,
    '3101A': 900,
    '51600': 1600,
    '51600S': 1600,
    'SS MINI': 1600,
    '62000': 2000,
    '62000-003': 2000,
    '62300': 2300,
    '62300XT': 2300,
    '52600': 2600,
    '52600S': 2600,
    'SSMINIXT': 2700
}

# whether or not the file is GPS-enabled (does not guarantee presence of GPS data in file)
GPS = {
    1: 'no',
    2: 'yes',
}

# bits per data word in radar array
BPS = {
    8: '8 unsigned',
    16: '16 unsigned',
    32: '32 signed'
}
